- Bulding Country recommender system based on languages spoken

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv(r"C:\Binjus Ds\ml\countries-languages.csv")

In [3]:
df

Country  \
0                        Afghanistan   
1                            Albania   
2                            Algeria   
3                            Andorra   
4                             Angola   
..                               ...   
193                          Vietnam   
194  Western Sahara (proposed state)   
195                            Yemen   
196                           Zambia   
197                         Zimbabwe   

                                      Languages Spoken  
0    Dari Persian, Pashtu (both official), other Tu...  
1       Albanian (Tosk is the official dialect), Greek  
2           Arabic (official), French, Berber dialects  
3    Catalán (official), French, Castilian, Portuguese  
4    Portuguese (official), Bantu and other African...  
..                                                 ...  
193  Vietnamese (official); English (increasingly f...  
194                  Hassaniya Arabic, Moroccan Arabic  
195                                             Arabic  
196  English (official); major vernaculars: Bemba, ...  
197  English (official), Shona, Ndebele (Sindebele)...  

[198 rows x 2 columns]

In [4]:
df.isnull().sum()

Country             0
Languages Spoken    0
dtype: int64

In [8]:
df.drop_duplicates()

Country  \
0                        Afghanistan   
1                            Albania   
2                            Algeria   
3                            Andorra   
4                             Angola   
..                               ...   
193                          Vietnam   
194  Western Sahara (proposed state)   
195                            Yemen   
196                           Zambia   
197                         Zimbabwe   

                                      Languages Spoken  
0    Dari Persian, Pashtu (both official), other Tu...  
1       Albanian (Tosk is the official dialect), Greek  
2           Arabic (official), French, Berber dialects  
3    Catalán (official), French, Castilian, Portuguese  
4    Portuguese (official), Bantu and other African...  
..                                                 ...  
193  Vietnamese (official); English (increasingly f...  
194                  Hassaniya Arabic, Moroccan Arabic  
195                                             Arabic  
196  English (official); major vernaculars: Bemba, ...  
197  English (official), Shona, Ndebele (Sindebele)...  

[198 rows x 2 columns]

In [9]:
df["Languages Spoken"].loc[0]

'Dari Persian, Pashtu (both official), other Turkic and minor languages'

### TF-IDF Matrix

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [11]:
tfidf= TfidfVectorizer(stop_words='english')

In [12]:
tfidf_matrix=tfidf.fit_transform(df["Languages Spoken"])

In [13]:
tfidf_matrix

<198x522 sparse matrix of type '<class 'numpy.float64'>'
	with 1307 stored elements in Compressed Sparse Row format>

In [14]:
from scipy import sparse

In [15]:
#converting sparse matrix to array
new_array=tfidf_matrix.toarray()
new_array

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [17]:
tfidf_matrix.shape

(198, 522)

- The vectorizer  has created 522 dimensional vector for the Languages_Spoken for every country

### Cosine similarity

In [18]:
from sklearn.metrics.pairwise import cosine_similarity

In [20]:
cosine_sim=cosine_similarity(tfidf_matrix,tfidf_matrix)

### Building Recommender

In [21]:
def content_recom(country,cosine_sim=cosine_sim,df=df):
    #get index of country matching input 
    idx=df[df["Country"]==country].index[0]
    # geting pairwise similarity scores of all countries against  that country and covert it into a list of tuples
    sim_scores=list(enumerate(cosine_sim[idx]))
    #sorting countries based on their similarity
    sim_scores=sorted(sim_scores,key= lambda x: x[1],reverse=True)
    # most similar 10 countries,ignore the first country
    sim_scores=sim_scores[1:6]
    #Getting the country indices
    country_indices=[i[0] for i in sim_scores]
    #returning top 5 countries which are similar to the input
    return df["Country"].iloc[country_indices]

In [25]:
content_recom("Algeria")

119         Morocco
152    Saudi Arabia
195           Yemen
85           Jordan
92           Kuwait
Name: Country, dtype: object

In [26]:
content_recom("Vietnam")

28                Cambodia
110       Marshall Islands
142                  Qatar
178    Trinidad and Tobago
24                  Brunei
Name: Country, dtype: object